src/python/pants/jvm/resolve/jvm_tool.py

```diff
@@ -0,0 +1,314 @@
+from dataclasses import dataclass
+from typing import ClassVar, Iterable, Sequence, cast
+
+from pants.backend.python.target_types import UnrecognizedResolveNamesError
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests, Workspace
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.rules import collect_rules, goal_rule, rule
+from pants.engine.unions import UnionMembership, union
+from pants.jvm.resolve.coursier_fetch import (
+    ArtifactRequirements,
+    Coordinate,
+    CoursierResolvedLockfile,
+)
+from pants.option.subsystem import Subsystem
+from pants.util.logging import LogLevel
+from pants.util.ordered_set import FrozenOrderedSet
+
+DEFAULT_TOOL_LOCKFILE = "<default>"
+NO_TOOL_LOCKFILE = "<none>"
+
+
+logger = logging.getLogger(__name__)
+
+
+class JvmToolBase(Subsystem):
+    """Base class for subsystems that configure a set of artifact requirements for a JVM tool."""
+
+    # Default version of the tool. (Subclasses must set.)
+    default_version: ClassVar[str]
+
+    # Default artifacts for the tool in GROUP:NAME format. The `--version` value will be used for the
+    # artifact version if it has not been specified for a particular requirement. (Subclasses must set.)
+    default_artifacts: ClassVar[Sequence[str]]
+
+    # Default extra requirements for the tool. (Subclasses do not need to override.)
+    default_extra_artifacts: ClassVar[Sequence[str]] = []
+
+    # Default resource for the tool's lockfile. (Subclasses must set.)
+    default_lockfile_resource: ClassVar[tuple[str, str]]
+
+    default_lockfile_url: ClassVar[str | None] = None
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        register(
+            "--version",
+            type=str,
+            advanced=True,
+            default=cls.default_version,
+            help=(
+                "Version string for the tool. This will be substituted for any unspecified version in "
+                "the --artifacts value."
+            ),
+        )
+        register(
+            "--artifacts",
+            type=list,
+            member_type=str,
+            advanced=True,
+            default=cls.default_artifacts,
+            help=(
+                "Artifact requirements for this tool using colon-separated form for the Maven coordinates. "
+                "The string %VERSION% version will be substituted with the --version value."
+            ),
+        )
+        register(
+            "--extra-artifacts",
+            type=list,
+            member_type=str,
+            advanced=True,
+            default=cls.default_extra_artifacts,
+            help="Any additional artifact requirement strings to use with the tool. This is useful if the "
+            "tool allows you to install plugins or if you need to constrain a dependency to "
+            "a certain version.",
+        )
```

Comment:
IMO, let's not add this until we know we need it... only some Python tools do, and the option ends up with a different meaning in each case. It's likely that we'd want to add specific options to subclasses with their own help, which could then override an empty method of the superclass to append them, for example.

Reply:
will remove